README.md
# Weatherdrop

![GitHub Workflow Status](<https://img.shields.io/github/workflow/status/bsord/weatherdrop/Create%20Release(s)?label=Docker%20Build>)
![GitHub top language](https://img.shields.io/github/languages/top/bsord/weatherdrop.svg)
![Docker Pulls](https://img.shields.io/docker/pulls/bsord/weatherdrop.svg)
![GitHub last commit](https://img.shields.io/github/last-commit/bsord/weatherdrop.svg)
![Lines of code](https://img.shields.io/tokei/lines/github/bsord/weatherdrop)
![License](https://img.shields.io/github/license/bsord/weatherdrop.svg?style=flat)

## Getting Started

First, copy env file :

```bash
cp .env-sample .env
```

and set OPEN_WEATHER_API_KEY

Then, run the development server:

```bash
npm install
npm run dev
# or with docker-compose
docker-compose up
```

Open [http://localhost:3000](http://localhost:3000) with your browser to see the result.

You can start editing the page by modifying `pages/index.tsx`. The page auto-updates as you edit the file.

[API routes](https://nextjs.org/docs/api-routes/introduction) can be accessed on [http://localhost:3000/api/hello](http://localhost:3000/api/hello). This endpoint can be edited in `pages/api/hello.tsx`.

The `pages/api` directory is mapped to `/api/*`. Files in this directory are treated as [API routes](https://nextjs.org/docs/api-routes/introduction) instead of React pages.


CHANGELOG.md
## [0.5.5](https://github.com/bsord/weatherdrop/compare/0.5.4...0.5.5) (2021-07-11)


### Bug Fixes

* Remove un-used giphy env var references ([2e81a0d](https://github.com/bsord/weatherdrop/commit/2e81a0d64b5e02d1d8b449dec400b4fe07d13793))



## [0.5.4](https://github.com/bsord/weatherdrop/compare/0.5.3...0.5.4) (2021-07-08)


### Bug Fixes

* Remove un-used random page ([d84fd7c](https://github.com/bsord/weatherdrop/commit/d84fd7c53ac4d44bd47be3da81e810342972bfda))



## [0.5.3](https://github.com/bsord/weatherdrop/compare/0.5.2...0.5.3) (2021-07-08)


### Bug Fixes

* crossOrigin warnings due to boolean when string expected ([942de68](https://github.com/bsord/weatherdrop/commit/942de683e05f4882d05dc0f9a85cef8020f8444c))



## [0.5.2](https://github.com/bsord/weatherdrop/compare/0.5.1...0.5.2) (2021-07-07)


### Bug Fixes

* **apis:** simplify api endpoint code, remove async, and add comments ([a98a369](https://github.com/bsord/weatherdrop/commit/a98a3696254776c6aff5b0c9152e029d9008bce1))



## [0.5.1](https://github.com/bsord/weatherdrop/compare/0.5.0...0.5.1) (2021-07-07)


### Bug Fixes

* rework client side api calls ([018ea0a](https://github.com/bsord/weatherdrop/commit/018ea0a326a0e6978ef915a838b7ddfeeaeb2464))
* Use larger width for gif display ([de9a4a9](https://github.com/bsord/weatherdrop/commit/de9a4a91fa0368238dd575ada7b08581f9ee8043))





pages/index.js
import Head from 'next/head'
import Link from 'next/link'
import styles from '../styles/Home.module.css'


export default function Home() {
  return (
    <div className={styles.container}>
      <Head>
        <title>WeatherDrop</title>
        <meta name="description" content="A page that reports the weather with memes and Gifs!" />
        <link rel="icon" href="/favicon.ico" />
        <link rel="stylesheet" href="/css/all.css" />
        <link rel="preconnect" href="https://fonts.googleapis.com" />
        <link rel="preconnect" href="https://fonts.gstatic.com" crossOrigin="true" />
        <link href="https://fonts.googleapis.com/css2?family=Roboto" rel="stylesheet" />
        <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />
        <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png" />
        <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png" />
        <link rel="manifest" href="/site.webmanifest" />
      </Head>

      <main className={styles.main}>
        <div className={styles.logo}>
          <i style={{ color:'#E5E5E5' }} className="fas fa-cloud-sun fa-9x"></i>
        </div>
        <h1 className={styles.title}>
          Welcome to <a href="https://nextjs.org">WeatherDrop</a>
        </h1>
        <h2 className={styles.description}>
          Your source for daily meme based weather forecasts
        </h2>
        <div className={styles.grid}>
          <Link href="/location">
            <a href="" className={styles.card}>
              <i style={{ float:"left", paddingRight:"10px" }} className="fas fa-map-marker-alt fa-2x"></i>
              <h2 style={{ color:"#FCA311" }}>My Forecast &rarr;</h2>
              <p style={{ color:'#E5E5E5' }}>Get the meme forecast based on your location</p>
            </a>
          </Link>
        </div>
      </main>

      <footer className={styles.footer}>
        <Link href="/">
          <a>WeatherDrop.io ©
            {' '}
            {new Date().getFullYear()} - v{process.env.NEXT_PUBLIC_APP_VERSION}
          </a>
        </Link>
      </footer>
    </div>
  )
}


pages/random.js
import Head from 'next/head'
import Link from 'next/link'
import styles from '../styles/Home.module.css'
import { GiphyFetch } from '@giphy/js-fetch-api'

export default function Random({gifUrl}) {
  return (
    <div className={styles.container}>
      <Head>
        <title>WeatherDrop - Random</title>
        <meta name="description" content="a Random meme forecast" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main className={styles.main}>
        <h1 className={styles.title}>
          Random Meme Forecast
        </h1>

        <p className={styles.description}>
          Click  <Link href="/random"><a>here</a></Link> to get a new random forecast
        </p>

        <div className={styles.grid}>
        {gifUrl && (
            <img src={gifUrl} alt="Meme forecast" />
          )}
        </div>
      </main>

      <footer className={styles.footer}>
        <a href="https://weatherdrop.io">
          WeatherDrop.io ©
          {' '}
          {new Date().getFullYear()}
        </a>
      </footer>
    </div>
  )
}

export async function getStaticProps() {

  // get Gif and pass to the page props
  const giphyApiKey = process.env.GIPHY_KEY || "" 
  let gifUrl = ""
  if(giphyApiKey != ""){
    const gf = new GiphyFetch(giphyApiKey)
    const { data: gifs } = await gf.search('weather', { sort: 'relevant', lang: 'es', limit: 1, type: 'gifs' })
    gifUrl = gifs[0].images.original.url
  } else {
    gifUrl = "https://media1.giphy.com/media/H7wajFPnZGdRWaQeu0/giphy.gif?cid=1995ed6coh4r816dv8hr3gw8f9p3clzvdsctlffxsvedomvr&rid=giphy.gif&ct=g"
  }
  
  return {
    props: { gifUrl},
  }
}

  

pages/api/gif.js
export default function handler(req, res) {
  let { weatherId, temp } = req.query
  let gifUrl = getRandomGif(weatherId, temp)
  return res.status(200).json({ gif: gifUrl })
}

function getRandomGif(weatherId, temp) {
  let gifUrl;
  if(temp <= 25){
      gifUrl = cold[Math.floor(Math.random() * cold.length)];
  }else if(temp >= 95){
    gifUrl = hot[Math.floor(Math.random() * hot.length)];
  }else {
    switch(true) {
      case weatherId >= 200 && weatherId <= 235:
        gifUrl = thunder[Math.floor(Math.random() * thunder.length)];
        break;
      case weatherId >= 300 && weatherId <= 325:
        gifUrl = drizzle[Math.floor(Math.random() * drizzle.length)];
        break;
      case weatherId >= 500 && weatherId <= 535:
        gifUrl = rain[Math.floor(Math.random() * rain.length)];
        break;
      case weatherId >= 600 && weatherId <= 625:
        gifUrl = snow[Math.floor(Math.random() * snow.length)];
        break;
      case weatherId >= 700 && weatherId <= 799:
        gifUrl = atmosphere[Math.floor(Math.random() * atmosphere.length)];
        break;
      case weatherId == 800:
        gifUrl = clear[Math.floor(Math.random() * clear.length)];
        break;
      case weatherId >= 801 && weatherId <= 804:
        gifUrl = clouds[Math.floor(Math.random() * clouds.length)];
        break;
      default:
        gifUrl = clear[Math.floor(Math.random() * clear.length)];
    }
  }
  return gifUrl;
}

let thunder = [
  "https://media.giphy.com/media/AFJAmPcCeH9RxiK9Y5/giphy.gif",
  "https://media.giphy.com/media/iLdNyukd3uxsk/giphy.gif",
  "https://media.giphy.com/media/3orif1fC5h9jcNX4go/giphy.gif",
  "https://media.giphy.com/media/l2Je1XYouqmuPhdi8/giphy.gif",
  "https://media.giphy.com/media/BgQDxZmWCEFiw/giphy.gif"
]

let drizzle = [
  "https://media.giphy.com/media/xT5LMAT8WK4InFirXG/giphy.gif",
  "https://media.giphy.com/media/za5xikuRr0OzK/giphy.gif",
  "https://media.giphy.com/media/xT9GEz2CeU9uaI2KZi/giphy.gif",
  "https://media.giphy.com/media/BgQDxZmWCEFiw/giphy.gif"
]

let rain = [
  "https://media.giphy.com/media/vSoXkKBccEQRq/giphy.gif",
  "https://media.giphy.com/media/l0HlPwMAzh13pcZ20/giphy.gif",
  "https://media.giphy.com/media/l2Je1XYouqmuPhdi8/giphy.gif",
  "https://media.giphy.com/media/BgQDxZmWCEFiw/giphy.gif",
  "https://media.giphy.com/media/l41m6dp4MjK9xTZ7y/giphy.gif",
  "https://media.giphy.com/media/xUNd9N08hpuyVHTUME/giphy.gif",
  "https://media.giphy.com/media/u0l83hl7HZYKBNeuGB/giphy.gif"
]

let snow = [
  "https://media.giphy.com/media/l2JIaYp6P3WT5Ybu0/giphy.gif",
  "https://media.giphy.com/media/SAC0wTRQYO2Y0/giphy.gif",
  "https://media.giphy.com/media/hkFgpYE8CRqog/giphy.gif",
  "https://media.giphy.com/media/5YDYrKUWKYeVW/giphy.gif",
  "https://media.giphy.com/media/10N782ExqDjCLK/giphy.gif",
  "https://media.giphy.com/media/JWegbsAWQS1YA/giphy.gif",
  "https://media.giphy.com/media/ugeAEY9a4dpYY/giphy.gif",
  "https://media.giphy.com/media/g2YvIlpgTMlck/giphy.gif",
  "https://media.giphy.com/media/Av0z9tdsBvFPLnnHz3/giphy.gif",
  "https://media.giphy.com/media/VRolEcq3Z4axa/giphy.gif",
  "https://media.giphy.com/media/10PrLKmZzVcK9a/giphy.gif",
  "https://media.giphy.com/media/EzhYp5ZQXomLm/giphy.gif"
]

let atmosphere = [
  "https://media.giphy.com/media/J9xKOmBdZlasbXzBjA/giphy.gif",
  "https://media.giphy.com/media/l0HlMURBbyUqF0XQI/giphy.gif"
]

let clear = [
  "https://media.giphy.com/media/Tj3gqingvdFZTokphI/giphy.gif",
  "https://media.giphy.com/media/xTEvYwzGarSU4J4FiM/giphy.gif",
  "https://media.giphy.com/media/LQthkp9wNfaMtmwG98/giphy.gif",
  "https://media.giphy.com/media/KA7zbi33B3a9i/giphy.gif",
  "https://media.giphy.com/media/JpGNLpC9jUdwxTrSEm/giphy.gif",
  "https://media.giphy.com/media/l08SqWKuJlyeLkYfLE/giphy.gif",
  "https://media.giphy.com/media/l08SqWKuJlyeLkYfLE/giphy.gif",
  "https://media.giphy.com/media/LnccQ4gwbYdd4O5PLR/giphy.gif",
  "https://media.giphy.com/media/5UqNjxKXGxDOhtR0hP/giphy.gif",
  "https://media.giphy.com/media/3orieMqnFY9nZjn9fy/giphy.gif",
  "https://media.giphy.com/media/g3WYJxZgCqLRu/giphy.gif",
  "https://media.giphy.com/media/xUPJUETissh9XI618s/giphy.gif",
  "https://media.giphy.com/media/mDSGaOdQxzdseSPdAi/giphy.gif",
  "https://media.giphy.com/media/JRgjhKV4UvgCpcue0q/giphy.gif",
  "https://media.giphy.com/media/lMDbNXxfQExri/giphy.gif"
]

let clouds = [
  "https://media.giphy.com/media/rrFcUcN3MFmta/giphy.gif",
  "https://media.giphy.com/media/26BGDQxDCZDFHW5Ne/giphy.gif",
  "https://media.giphy.com/media/HknSLLEbzZCoM/giphy.gif",
  "https://media.giphy.com/media/l0HlQdk8kI9KIOjBe/giphy.gif",
  "https://media.giphy.com/media/3orieMAp8v8mtnVvJm/giphy.gif",
  "https://media.giphy.com/media/McZdU5M5jpkmk/giphy.gif"
]

let cold = [
  "https://media.giphy.com/media/nVGyjauCDpAFq/giphy.gif",
  "https://media.giphy.com/media/KFUx0Rtz7p0HTzbJ7x/giphy.gif",
  "https://media.giphy.com/media/KaW6fNYZf6eSk/giphy.gif",
  "https://media.giphy.com/media/l0NwI6LbXx4sWQCmk/giphy.gif",
  "https://media.giphy.com/media/xUOwGoV4fmBd0fiyGI/giphy.gif",
  "https://media.giphy.com/media/l0NwtuI3FrBQ7xk4g/giphy.gif",
  "https://media.giphy.com/media/4L69OKSHdCx7a/giphy.gif",
  "https://media.giphy.com/media/3o6fDMaNGxz2rblFQs/giphy.gif",
  "https://media.giphy.com/media/xTeV7qibUsOYA3mmYw/giphy.gif",
  "https://media.giphy.com/media/3o7ZeCHGCq8vJgj4GY/giphy.gif",
  "https://media.giphy.com/media/3o7bugMEMJx8235gac/giphy.gif",
  "https://media.giphy.com/media/s4Bi420mMDRBK/giphy.gif",
  "https://media.giphy.com/media/l49FiEAsjmPdoYhVK/giphy.gif"
]

let hot = [
  "https://media.giphy.com/media/l2Je4PGbVcHDsQhnG/giphy.gif",
  "https://media.giphy.com/media/LjpIl7WfncQgw/giphy.gif",
  "https://media.giphy.com/media/LgULLE6V45LfG/giphy.gif",
  "https://media.giphy.com/media/xT5LMAUnrvWDGvwLQY/giphy.gif",
  "https://media.giphy.com/media/Y4h6SX80viZIKu76d2/giphy.gif",
  "https://media.giphy.com/media/nrXif9YExO9EI/giphy.gif",
  "https://media.giphy.com/media/hkik4ac9sSqaY/giphy.gif",
  "https://media.giphy.com/media/l2Je2UdtBFPSxxIIw/giphy.gif",
  "https://media.giphy.com/media/11oRLY4FRk2s36/giphy.gif",
  "https://media.giphy.com/media/l0HlMURBbyUqF0XQI/giphy.gif",
  "https://media.giphy.com/media/NTur7XlVDUdqM/giphy.gif"
]

pages/location.js
import Head from 'next/head'
import Link from 'next/link'
import styles from '../styles/Location.module.css'
import React, { useState,useEffect } from 'react';

export default function Location() {
  const [city, setCity] = useState(null);
  const [temp, setTemp] = useState(null);
  const [feels, setFeels] = useState(null);
  const [low, setLow] = useState(null);
  const [high, setHigh] = useState(null);
  const [weather, setWeather] = useState(null);
  const [weatherGif, setWeatherGif] = useState(null);

  const getLocation = () => {
    if (!navigator.geolocation) {
      console.log("This browser does not support Geolocation")
    } else {
      navigator.geolocation.getCurrentPosition((position) => {
        //console.log(position)
        fetchSummaryJSON(position.coords.latitude, position.coords.longitude)
      }, () => {
        console.log("Unable to retrieve your location");
      });
    }
  }

  function getWeatherGif(id, temp) {
    let apiUrl = `/api/gif?weatherId=${id}&temp=${temp}`;
    fetch(apiUrl)
      .then(response => response.json())
      .then(data => setWeatherGif(data.gif));
  };

  function fetchSummaryJSON(lat, lng) {
    let apiUrl = `/api/summary?lat=${lat}&long=${lng}`;
    fetch(apiUrl)
      .then(response => response.json())
      .then(forecast => {
        getWeatherGif(forecast.weather_status_id, forecast.temp_actual)
        setCity(forecast.city)
        setWeather(forecast.weather_status)
        setTemp(forecast.temp_actual)
        setFeels(forecast.temp_feels)
        setLow(forecast.temp_low)
        setHigh(forecast.temp_high)
      });

  }
  
  useEffect(()=>{
    getLocation()
  }, []);

  return (
    <div className={styles.container} >
      <Head>
        <title>WeatherDrop - Location</title>
        <meta name="description" content="a Random meme forecast" />
        <link rel="icon" href="/favicon.ico" />
        <link rel="stylesheet" href="/css/all.css" />
        <link rel="preconnect" href="https://fonts.googleapis.com" />
        <link rel="preconnect" href="https://fonts.gstatic.com" crossOrigin="true" />
        <link href="https://fonts.googleapis.com/css2?family=Roboto" rel="stylesheet" />
        <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />
        <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png" />
        <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png" />
        <link rel="manifest" href="/site.webmanifest" />
      </Head>
      <main className={styles.main}>
        <div className={styles.card}>
          {
              weather && weatherGif &&
              <div className={styles.logo}>
                <img src={weatherGif} alt={weather + "-gif"} width="75%" height="75%"/>
              </div>
          }
          {
            city &&
            <h1 className={styles.title}>
              {city}
            </h1>
          }
          {
            weather &&
            <h3 className={styles.description}>
               <i style={{ paddingRight:"5px" }} className="fas fa-thermometer-three-quarters"/> {temp}&deg; F &nbsp;&nbsp; | &nbsp;&nbsp; <i style={{ paddingRight:"5px" }} className="fas fa-sun"/> {weather}
            </h3>
          }
          {
            temp && feels &&
            <p className={styles.description}>
              <i className="fas fa-sun"/> Feels Like: {feels}&deg; F
            </p>
          }
          {
            low && high &&
            <div>
              <p className={styles.description}>
                <i className="fas fa-level-down-alt"/> Low: {low}&deg; F &nbsp;&nbsp; | &nbsp;&nbsp; <i className="fas fa-level-up-alt"/> High: {high}&deg; F
              </p>
            </div>
          }
        </div>
      </main>

      <footer className={styles.footer}>
        <Link href="/">
          <a>WeatherDrop.io ©
            {' '}
            {new Date().getFullYear()}
          </a>
        </Link>
      </footer>
    </div>
  )
}

pages/api/summary.js
//
// API Docs: https://openweathermap.org/current
//
let apiKey;
if(process.env.OPEN_WEATHER_API_KEY){
  apiKey = process.env.OPEN_WEATHER_API_KEY;
}else{
  throw new Error("OpenWeather API Key not found!");
}

export default function handler(req, res) {
  let { lat, long } = req.query
  let apiUrl = `https://api.openweathermap.org/data/2.5/weather?lat=${lat}&lon=${long}&appid=${apiKey}&units=imperial`;
  return fetch(apiUrl)
    .then(response => response.json())
    .then(forecast => {
      if('cod' in forecast && forecast.cod == '200'){
        res.status(200).json({
          city: forecast.name,
          city_id: forecast.id,
          temp_actual: forecast.main.temp.toFixed(1),
          temp_feels: forecast.main.feels_like.toFixed(1),
          temp_low: forecast.main.temp_min.toFixed(1),
          temp_high: forecast.main.temp_max.toFixed(1),
          weather_status: forecast.weather[0].main,
          weather_status_desc: forecast.weather[0].description,
          weather_status_id: forecast.weather[0].id,
        })
      } else { throw new Error("Failed to get OpenWeather data") }
    });
}


pages/api/forecast.ts
import type { NextApiRequest, NextApiResponse } from 'next'

type Data = {
  city: string,
  temp_actual: number,
  temp_feels: number,
  temp_min: number,
  temp_max: number,
  weather_status: string,
  weather_desc: string,
  typescript: string
}

type Error = {
  message: string
}


export default function handler(req: NextApiRequest, res: NextApiResponse<Data | Error>) {
  let apiKey;
  if(process.env.OPEN_WEATHER_API_KEY !== undefined || process.env.OPEN_WEATHER_API_KEY !== "undefined"){
    apiKey = process.env.OPEN_WEATHER_API_KEY;
  }else{
    new Error("OpenWeather API Key not found!");
  }
  let { lat, long } = req.query
  let apiUrl = `https://api.openweathermap.org/data/2.5/weather?lat=${lat}&lon=${long}&appid=${apiKey}&units=imperial`;

  async function fetchForecastJSON() {
    const response = await fetch(apiUrl);
    const forecast = await response.json();
    return forecast;
  }

  try {
    fetchForecastJSON().then(forecast => {
      res.status(200).json({
        city: forecast.name,
        temp_actual: forecast.main.temp,
        temp_feels: forecast.main.feels_like,
        temp_min: forecast.main.temp_min,
        temp_max: forecast.main.temp_max,
        weather_status: forecast.weather[0].main,
        weather_desc: forecast.weather[0].description,
        typescript: "sucks"
      })
    });
  } catch(err) {
    res.status(500).json({
      message: "There was an error with the request"
      //be careful about exposing backend errors
    })
    console.log(err)
  }
}
